File: src-tauri/src/db/prompts.rs
// db/prompts.rs — Prompts 仓储。所有 SQL 在此文件，commands 只做参数转发。
use rusqlite::{params, Connection, Row};

use crate::error::{AppError, AppResult};
use crate::models::prompt::{Prompt, PromptDraft, SortMode};
use super::now_ms;

fn map_row(r: &Row<'_>) -> rusqlite::Result<Prompt> {
    Ok(Prompt {
        id: r.get(0)?,
        title: r.get(1)?,
        content: r.get(2)?,
        folder_id: r.get(3)?,
        tag_ids: vec![], // 单独查询
        is_favorite: r.get::<_, i64>(4)? != 0,
        is_pinned: r.get::<_, i64>(5)? != 0,
        use_count: r.get(6)?,
        last_used_at: r.get(7)?,
        created_at: r.get(8)?,
        updated_at: r.get(9)?,
    })
}

const SELECT_COLS: &str = "id, title, content, folder_id, is_favorite, is_pinned, \
    use_count, last_used_at, created_at, updated_at";

fn order_clause(sort: SortMode) -> &'static str {
    match sort {
        SortMode::RecentUsed => {
            "is_pinned DESC, last_used_at IS NULL, last_used_at DESC, updated_at DESC"
        }
        SortMode::Created => "is_pinned DESC, created_at DESC",
        SortMode::Updated => "is_pinned DESC, updated_at DESC",
        SortMode::Title => "is_pinned DESC, title COLLATE NOCASE ASC",
    }
}

fn fetch_tag_ids(conn: &Connection, prompt_id: i64) -> AppResult<Vec<i64>> {
    let mut stmt = conn
        .prepare("SELECT tag_id FROM prompt_tags WHERE prompt_id = ?1 ORDER BY tag_id")
        .map_err(|e| AppError::Db(e.to_string()))?;
    let rows = stmt
        .query_map(params![prompt_id], |r| r.get::<_, i64>(0))
        .map_err(|e| AppError::Db(e.to_string()))?;
    let mut out = Vec::new();
    for r in rows {
        out.push(r.map_err(|e| AppError::Db(e.to_string()))?);
    }
    Ok(out)
}

pub fn list(conn: &Connection, sort: SortMode) -> AppResult<Vec<Prompt>> {
    let sql = format!(
        "SELECT {SELECT_COLS} FROM prompts ORDER BY {}",
        order_clause(sort)
    );
    let mut stmt = conn.prepare(&sql).map_err(|e| AppError::Db(e.to_string()))?;
    let mut prompts: Vec<Prompt> = stmt
        .query_map([], map_row)
        .map_err(|e| AppError::Db(e.to_string()))?
        .collect::<rusqlite::Result<_>>()
        .map_err(|e| AppError::Db(e.to_string()))?;
    for p in prompts.iter_mut() {
        p.tag_ids = fetch_tag_ids(conn, p.id)?;
    }
    Ok(prompts)
}

pub fn get(conn: &Connection, id: i64) -> AppResult<Prompt> {
    let mut stmt = conn
        .prepare(&format!("SELECT {SELECT_COLS} FROM prompts WHERE id = ?1"))
        .map_err(|e| AppError::Db(e.to_string()))?;
    let mut p = stmt
        .query_row(params![id], map_row)
        .map_err(|e| match e {
            rusqlite::Error::QueryReturnedNoRows => {
                AppError::NotFound(format!("prompt {id}"))
            }
            other => AppError::Db(other.to_string()),
        })?;
    p.tag_ids = fetch_tag_ids(conn, id)?;
    Ok(p)
}

fn replace_tags(conn: &Connection, prompt_id: i64, tag_ids: &[i64]) -> AppResult<()> {
    conn.execute("DELETE FROM prompt_tags WHERE prompt_id = ?1", params![prompt_id])
        .map_err(|e| AppError::Db(e.to_string()))?;
    let mut stmt = conn
        .prepare("INSERT INTO prompt_tags (prompt_id, tag_id) VALUES (?1, ?2)")
        .map_err(|e| AppError::Db(e.to_string()))?;
    for tid in tag_ids {
        stmt.execute(params![prompt_id, tid])
            .map_err(|e| AppError::Db(e.to_string()))?;
    }
    Ok(())
}

pub fn create(conn: &mut Connection, draft: PromptDraft) -> AppResult<Prompt> {
    if draft.title.trim().is_empty() {
        return Err(AppError::InvalidInput("title is empty".into()));
    }
    let now = now_ms();
    let tx = conn.transaction().map_err(|e| AppError::Db(e.to_string()))?;
    tx.execute(
        "INSERT INTO prompts (title, content, folder_id, is_favorite, is_pinned, \
         use_count, created_at, updated_at) VALUES (?1, ?2, ?3, 0, 0, 0, ?4, ?4)",
        params![draft.title, draft.content, draft.folder_id, now],
    )
    .map_err(|e| AppError::Db(e.to_string()))?;
    let id = tx.last_insert_rowid();
    replace_tags(&tx, id, &draft.tag_ids)?;
    tx.commit().map_err(|e| AppError::Db(e.to_string()))?;
    get(conn, id)
}

pub fn update(conn: &mut Connection, id: i64, draft: PromptDraft) -> AppResult<Prompt> {
    if draft.title.trim().is_empty() {
        return Err(AppError::InvalidInput("title is empty".into()));
    }
    let tx = conn.transaction().map_err(|e| AppError::Db(e.to_string()))?;
    let n = tx
        .execute(
            "UPDATE prompts SET title=?1, content=?2, folder_id=?3, updated_at=?4 \
             WHERE id=?5",
            params![draft.title, draft.content, draft.folder_id, now_ms(), id],
        )
        .map_err(|e| AppError::Db(e.to_string()))?;
    if n == 0 {
        return Err(AppError::NotFound(format!("prompt {id}")));
    }
    replace_tags(&tx, id, &draft.tag_ids)?;
    tx.commit().map_err(|e| AppError::Db(e.to_string()))?;
    get(conn, id)
}

pub fn delete(conn: &Connection, id: i64) -> AppResult<()> {
    let n = conn
        .execute("DELETE FROM prompts WHERE id = ?1", params![id])
        .map_err(|e| AppError::Db(e.to_string()))?;
    if n == 0 {
        return Err(AppError::NotFound(format!("prompt {id}")));
    }
    Ok(())
}

pub fn toggle_favorite(conn: &Connection, id: i64) -> AppResult<Prompt> {
    let n = conn
        .execute(
            "UPDATE prompts SET is_favorite = 1 - is_favorite, updated_at = ?1 \
             WHERE id = ?2",
            params![now_ms(), id],
        )
        .map_err(|e| AppError::Db(e.to_string()))?;
    if n == 0 {
        return Err(AppError::NotFound(format!("prompt {id}")));
    }
    get(conn, id)
}

pub fn toggle_pin(conn: &Connection, id: i64) -> AppResult<Prompt> {
    let n = conn
        .execute(
            "UPDATE prompts SET is_pinned = 1 - is_pinned, updated_at = ?1 \
             WHERE id = ?2",
            params![now_ms(), id],
        )
        .map_err(|e| AppError::Db(e.to_string()))?;
    if n == 0 {
        return Err(AppError::NotFound(format!("prompt {id}")));
    }
    get(conn, id)
}

pub fn record_use(conn: &Connection, id: i64) -> AppResult<()> {
    conn.execute(
        "UPDATE prompts SET use_count = use_count + 1, last_used_at = ?1 WHERE id = ?2",
        params![now_ms(), id],
    )
    .map_err(|e| AppError::Db(e.to_string()))?;
    Ok(())
}

File: app/db/repositories/prompts.py
"""Prompt CRUD + ordering."""
from __future__ import annotations

from collections import defaultdict
from collections.abc import Iterable

from app.db.connection import get_conn, transaction
from app.models import Prompt, SortMode, now_ms

_ORDER_SQL: dict[SortMode, str] = {
    SortMode.RECENT_USED: "is_pinned DESC, COALESCE(last_used_at, 0) DESC, updated_at DESC",
    SortMode.CREATED: "is_pinned DESC, created_at DESC",
    SortMode.UPDATED: "is_pinned DESC, updated_at DESC",
    SortMode.TITLE: "is_pinned DESC, title COLLATE NOCASE ASC",
}


def _load_tag_map(prompt_ids: Iterable[int]) -> dict[int, list[int]]:
    ids = list(prompt_ids)
    if not ids:
        return {}
    placeholders = ",".join("?" * len(ids))
    rows = get_conn().execute(
        f"SELECT prompt_id, tag_id FROM prompt_tags WHERE prompt_id IN ({placeholders})",
        ids,
    ).fetchall()
    out: dict[int, list[int]] = defaultdict(list)
    for r in rows:
        out[r["prompt_id"]].append(r["tag_id"])
    return out


def list_all(sort_mode: SortMode = SortMode.RECENT_USED) -> list[Prompt]:
    order = _ORDER_SQL[sort_mode]
    rows = get_conn().execute(f"SELECT * FROM prompts ORDER BY {order}").fetchall()
    tags = _load_tag_map(r["id"] for r in rows)
    return [Prompt.from_row(r, tags.get(r["id"], [])) for r in rows]


def get(prompt_id: int) -> Prompt | None:
    row = get_conn().execute("SELECT * FROM prompts WHERE id = ?", (prompt_id,)).fetchone()
    if row is None:
        return None
    tags = _load_tag_map([prompt_id]).get(prompt_id, [])
    return Prompt.from_row(row, tags)


def create(*, title: str, content: str, folder_id: int | None = None, tag_ids: Iterable[int] = ()) -> int:
    ts = now_ms()
    with transaction() as conn:
        cur = conn.execute(
            "INSERT INTO prompts(title, content, folder_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            (title, content, folder_id, ts, ts),
        )
        prompt_id = cur.lastrowid
        _replace_tags(conn, prompt_id, tag_ids)
        return prompt_id


def update(
    prompt_id: int,
    *,
    title: str | None = None,
    content: str | None = None,
    folder_id: int | None = None,
    tag_ids: Iterable[int] | None = None,
    clear_folder: bool = False,
) -> None:
    fields: list[str] = []
    params: list[object] = []
    if title is not None:
        fields.append("title = ?")
        params.append(title)
    if content is not None:
        fields.append("content = ?")
        params.append(content)
    if folder_id is not None or clear_folder:
        fields.append("folder_id = ?")
        params.append(None if clear_folder else folder_id)
    fields.append("updated_at = ?")
    params.append(now_ms())
    params.append(prompt_id)

    with transaction() as conn:
        if fields:
            conn.execute(f"UPDATE prompts SET {', '.join(fields)} WHERE id = ?", params)
        if tag_ids is not None:
            _replace_tags(conn, prompt_id, tag_ids)


def delete(prompt_id: int) -> None:
    with transaction() as conn:
        conn.execute("DELETE FROM prompts WHERE id = ?", (prompt_id,))


def toggle_favorite(prompt_id: int) -> bool:
    with transaction() as conn:
        conn.execute("UPDATE prompts SET is_favorite = 1 - is_favorite WHERE id = ?", (prompt_id,))
        row = conn.execute("SELECT is_favorite FROM prompts WHERE id = ?", (prompt_id,)).fetchone()
        return bool(row["is_favorite"]) if row else False


def toggle_pin(prompt_id: int) -> bool:
    with transaction() as conn:
        conn.execute("UPDATE prompts SET is_pinned = 1 - is_pinned WHERE id = ?", (prompt_id,))
        row = conn.execute("SELECT is_pinned FROM prompts WHERE id = ?", (prompt_id,)).fetchone()
        return bool(row["is_pinned"]) if row else False


def record_use(prompt_id: int) -> None:
    ts = now_ms()
    with transaction() as conn:
        conn.execute(
            "UPDATE prompts SET use_count = use_count + 1, last_used_at = ? WHERE id = ?",
            (ts, prompt_id),
        )


def _replace_tags(conn, prompt_id: int, tag_ids: Iterable[int]) -> None:
    conn.execute("DELETE FROM prompt_tags WHERE prompt_id = ?", (prompt_id,))
    rows = [(prompt_id, tid) for tid in tag_ids]
    if rows:
        conn.executemany("INSERT INTO prompt_tags(prompt_id, tag_id) VALUES (?, ?)", rows)
